### PR TITLE
[REEF-1160] Deprecate O.A.R.Driver.ContextConfiguration and add O.A.R…

### PR DESCRIPTION
**Pull request by afchung**
….Common.Context.ContextConfiguration

This addressed the issue by
  * Deprecated O.A.R.Driver.ContextConfiguration (and options) and add O.A.R.Common.Context.ContextConfiguration (and options).
  * Added tests for currently the only functionality of ContextConfigurations -- allowing submission of Context IDs.
  * Deprecated default context handlers and created new classes for them in O.A.R.Common.

JIRA:
  [REEF-1160](https://issues.apache.org/jira/browse/REEF-1160)

### CODE REVIEW

**Comment by afchung** [comment]
@jwang98052 could you please have a look when you have time? This will be blocking REEF-1063, REEF-1062, and REEF-1144. Thanks!

File: lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    public class DefaultContextMessageSource : IContextMessageSource
+    {
+        public Optional<ContextMessage> Message
+        {
+            get
+            {
+                return Optional<ContextMessage>.Empty();
+            }
+
+            set
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs]: Why is there an empty `set` here?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs [hunk 2]
```diff
@@ -36,17 +38,21 @@ internal sealed class ContextManager : IDisposable
...
         private readonly RootContextLauncher _rootContextLauncher;
         private readonly object _contextLock = new object();
+        private readonly AvroConfigurationSerializer _serializer;
         private ContextRuntime _topContext = null;
 
         [Inject]
         private ContextManager(
             HeartBeatManager heartBeatManager,
-            EvaluatorSettings evaluatorSetting)
+            EvaluatorSettings evaluatorSetting,
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs]: typo? `evaluatorSetting` -> `evaluatorSettings`

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 3]
```diff
@@ -93,14 +90,35 @@ internal sealed class ContextRuntime
         }
 
         /// <summary>
+        /// Create a new ContextRuntime.
+        /// </summary>
+        /// <param name="serviceInjector"></param>
+        /// <param name="contextConfiguration">the Configuration for this context.</param>
+        /// <param name="parentContext"></param>
+        public ContextRuntime(
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Does this have to be `public`?

File: lang/cs/Org.Apache.REEF.Driver/Context/ContextConfiguration.cs [hunk 4]
```diff
@@ -30,6 +30,7 @@
 namespace Org.Apache.REEF.Driver.Context
 {
     [ClientSide]
+    [Obsolete("Obsoleted in 0.14, please use Org.Apache.REEF.Common.ContextConfiguration instead.")]
     public sealed class ContextConfiguration : ConfigurationModuleBuilder
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Driver/Context/ContextConfiguration.cs]: Can you also use the new named parameters instead?

**Comment by markusweimer** [comment]
I had another look. The implementations of contexts on the Evaluator side is a stop-gap until we do the proper rewrite, right?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 3 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: `ContextRuntime` is an `internal` class.

File: lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs (hunk 1 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs]: `IContextMessageSource` requires both `get` and `set` to be implemented, which I think is a bad idea, but this should be addressed in a separate PR.

**Comment by afchung** [comment]
@markusweimer yes, this is just a stop-gap such that we support both the new and the old `ContextConfiguration` s in the meantime. The old `ContextConfiguration` is really just a container for `ContextId` at this point though...

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs (hunk 2 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs]: Good catch.

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Tang.Util;
+using Org.Apache.REEF.Utilities.Attributes;
+
+[module: SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1401:FieldsMustBePrivate", Justification = "static field, typical usage in configurations")]
+
+namespace Org.Apache.REEF.Common.Context
+{
+    [ClientSide]
+    public sealed class ContextConfiguration : ConfigurationModuleBuilder
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs]: You might use this opportunity to rename it as ContextConfigurationModule to distinguish it from another ContextConfiguration.

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Utilities.Attributes;
+
+namespace Org.Apache.REEF.Common.Context
+{
+    /// <summary>
+    ///  Configuration parameters for ContextConfiguration module.
+    /// </summary>
+    [ClientSide]
+    public sealed class ContextConfigurationOptions : ConfigurationModuleBuilder
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs]: Why inherit from ConfigurationModuleBuilder? Should clean up.

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs (hunk 5 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs]: I would like to keep the parity between C# and Java. Java uses `ContextConfiguration`. The other `ContextConfiguration` will eventually be deleted.

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs (hunk 6 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs]: Not sure, it was there from before. I'll clean this up.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 7]
```diff
@@ -55,21 +55,18 @@ internal sealed class ContextRuntime
...
                 Optional<ContextRuntime> parentContext)
         {
-            var config = contextConfiguration as ContextConfiguration;
-            if (config == null)
-            {
-                Utilities.Diagnostics.Exceptions.Throw(
-                    new ArgumentException("contextConfiguration is not of type ContextConfiguration"), LOGGER);
-            }
-            _contextLifeCycle = new ContextLifeCycle(config.Id);
+            _contextLifeCycle = new ContextLifeCycle(id);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: You make it injectable, but still use new here.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 8]
```diff
@@ -93,14 +90,35 @@ internal sealed class ContextRuntime
...
+        /// <param name="id">the ID of the context.</param>
+        /// <param name="serviceInjector">the serviceInjector to be used.</param>
         /// <param name="contextConfiguration"> the Configuration for this context.</param>
+        [Obsolete("Deprecated in 0.14, will be removed.")]
         public ContextRuntime(
+            string id,
             IInjector serviceInjector,
             IConfiguration contextConfiguration)
-            : this(serviceInjector, contextConfiguration, Optional<ContextRuntime>.Empty())
+            : this(id, serviceInjector, contextConfiguration, Optional<ContextRuntime>.Empty())
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: If it is going to be obsolete, why brother to change it to add id parameter here?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 9]
```diff
@@ -141,8 +170,9 @@ public ContextRuntime SpawnChildContext(IConfiguration contextConfiguration, ICo
                 }
                 try
                 {
-                    IInjector childServiceInjector = _serviceInjector.ForkInjector(serviceConfiguration);
-                    var childContext = new ContextRuntime(childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(this));
+                    var childServiceInjector = _serviceInjector.ForkInjector(serviceConfiguration);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Was _ServiceInjector not created using serviceConfiguration?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs [hunk 10]
```diff
@@ -115,6 +133,17 @@ public Optional<ContextRuntime> ParentContext
...
 
+        private static ContextRuntime GetChildContextRuntime(IInjector childServiceInjector, IConfiguration contextConfiguration, ContextRuntime parentRuntime)
+        {
+            var actualContextConfiguration = contextConfiguration as ContextConfiguration;
+            if (actualContextConfiguration == null)
+            {
+                return new ContextRuntime(childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(parentRuntime));
+            }
+            
+            return new ContextRuntime(actualContextConfiguration.Id, childServiceInjector, contextConfiguration, Optional<ContextRuntime>.Of(parentRuntime));
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: still use obsoleted ContextRuntime?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 8 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Because we still need to support the old way to instantiate things in the meantime (i.e. using `ContextConfiguration.Id`.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 7 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: We are having two paths, one is via injection, and one is via the to-be-removed path. This is the to-be-removed path.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 9 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: this is the child service configuration, they are different.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 10 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Yes, it will be Obsoleted in the 0.14 release, but we cannot yet remove it, so we have to support both paths prior to the 0.15 release.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs [hunk 11]
```diff
@@ -35,35 +35,33 @@ internal sealed class RootContextLauncher
...
-            set { _rootContextConfiguration = value; }
-        }
+        public string Id { get; private set; }
 
         internal ContextRuntime GetRootContext()
         {
             if (_rootContext == null)
             {
-                _rootContext = GetRootContext(_rootServiceInjector, _rootContextConfiguration);
+                _rootContext = new ContextRuntime(Id, _rootServiceInjector, _rootContextConfiguration);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs]: This is still obsolete one. Should you put TODO and add a Jira for it?

File: lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSimpleContext.cs [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            CleanUp(testFolder);
+            TestRun(configuration, typeof(TestContextHandlers), 1, "testSimpleContext", "local", testFolder);
+            ValidateSuccessForLocalRuntime(1, testFolder: testFolder);
+            ValidateMessageSuccessfullyLogged(ValidationMessage, testFolder);
+            CleanUp(testFolder);
+        }
+
+        private static IConfiguration MergeAssembliesConfiguration(IConfiguration configuration)
+        {
+            return TangFactory.GetTang().NewConfigurationBuilder(configuration)
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSimpleContext.cs]: This should be not needed. You can remove it and have a try.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Then you might want to name serviceConfiguration as childServiceConfiguration to make it clear.

**Comment by jwang98052** [comment]
I completed my pass. 

Jira mentioned it is file moving. But looks like you have changes more than that in this PR. Could you describe it in either Jira or this PR description to make it clear.

File: lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs]: Agreed. I filled [REEF-1166](https://issues.apache.org/jira/browse/REEF-1166) for it.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 9 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: Done

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs (hunk 11 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/RootContextLauncher.cs]: Filed REEF-1167.

**Comment by afchung** [comment]
@jwang98052 @markusweimer I've addressed your comments, modified the description JIRA, and opened REEF-1167 for the removal phase. Please have another look, thanks for the reviews!

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+using Org.Apache.REEF.Common.Events;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Attributes;
+
+namespace Org.Apache.REEF.Common.Context
+{
+    /// <summary>
+    ///  Configuration parameters for ContextConfiguration module.
+    /// </summary>
+    /// TODO[JIRA REEF-1167]: Remove this class.
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs]: Isn't it the other copy of this class we'd like to remove? Also, it would be good for old a new code paths to share their named parameters.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs (hunk 3 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextRuntime.cs]: fair enough

File: lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs (hunk 13 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/ContextConfigurationOptions.cs]: Whoops, added to the wrong class. Fixed.

File: lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs (hunk 1 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs]: That PR just went in. We should remove this setter.

Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Context/Defaults/DefaultContextMessageSource.cs]: Sounds good. I'll make a merge commit and fix this.

**Comment by afchung** [comment]
@jwang98052 @markusweimer Could you please have another look when available? Thanks!

**Comment by markusweimer** [comment]
Looks good to me. @jwang98052?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs [hunk 14]
```diff
@@ -276,8 +282,18 @@ private void AddContext(AddContextProto addContextProto)
                         currentTopContext.Id));
                     Utilities.Diagnostics.Exceptions.Throw(e, LOGGER);
                 }
-                string contextConfigString = addContextProto.context_configuration;
-                var contextConfiguration = new ContextConfiguration(contextConfigString);
+
+                IConfiguration contextConfiguration;
+                try
+                {
+                    contextConfiguration = _serializer.FromString(addContextProto.context_configuration);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Context/ContextManager.cs]: In which situation, this deserialization may get exception?

**Comment by jwang98052** [comment]
I will test and merge

**Comment by afchung** [comment]
@jwang98052 after reading your earlier question, I've realized a mistake I've made in the PR, but I didn't think you would merge this since I have not yet answered the question. Sorry! I've created a new item (REEF-1172) to address the mistake, and will create a new PR soon. Please have a look at it. Thanks!
